// tradehub-api-js/build/main/lib/clients/ws.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WsClient = exports.ClientEvent = exports.MarketEvent = void 0;
var events_1 = require("events");
var config_1 = require("../config");
var isomorphic_ws_1 = __importDefault(require("isomorphic-ws"));
/**
 * @deprecated use WSConnector
 */
var MarketEvent;
(function (MarketEvent) {
    MarketEvent["SUB_MARKET_STATS"] = "market:sub:market_stats";
})(MarketEvent = exports.MarketEvent || (exports.MarketEvent = {}));
/**
 * @deprecated use WSConnector
 */
var ClientEvent;
(function (ClientEvent) {
    ClientEvent["Connected"] = "connected";
    ClientEvent["Disconnected"] = "disconnected";
    ClientEvent["Message"] = "message";
})(ClientEvent = exports.ClientEvent || (exports.ClientEvent = {}));
/**
 * @deprecated use WSConnector
 */
var WsClient = /** @class */ (function (_super) {
    __extends(WsClient, _super);
    // private channelIdToId: Map<string, string> = new Map()
    function WsClient(network) {
        var _this = _super.call(this) || this;
        // this.socket = this.newWebSocket(getNetwork(network).WS_URL)
        _this.baseUrl = config_1.getNetwork(network).WS_URL;
        return _this;
    }
    // private newWebSocket(): WebSocket {
    //   this.socket = new WebSocket(this.baseUrl)
    //   this.socket.onopen = () => {
    //     console.log('ws connected')
    //     this.emit(ClientEvent.Connect)
    //   }
    //   this.socket.onclose = () => this.emit(ClientEvent.Disconnect)
    // this.socket.onmessage = (message) => {
    //   console.log('message', message)
    //   const data = JSON.parse(message.data)
    //   const {channel, result} = data
    //   let {id} = data
    //   if (id) {
    //     const action = id.split(':')[1]
    //     if (action === 'sub') {
    //       const [channelId] = result
    //       this.channelIdToId.set(channelId, id)
    //       id += ':ack'
    //     }
    //   } else if (channel) {
    //     id = this.channelIdToId.get(channel)
    //   }
    //   else {
    //     console.error('Unknown event:', data)
    //     return
    //   }
    //   const [clientName] = id.split(':')
    //   const event = {...data, id}
    //   const client = this[clientName]
    //   if (client) {
    //     client.ws.handleEvent(event)
    //   } else {
    //     console.error('Unknown client:', clientName)
    //   }
    // }
    // }
    WsClient.prototype.connect = function () {
        var _this = this;
        this.socket = new isomorphic_ws_1.default(this.baseUrl);
        this.socket.onopen = function () {
            _this.emit(ClientEvent.Connected);
        };
        this.socket.onclose = function () { return _this.emit(ClientEvent.Disconnected); };
        this.socket.onmessage = function (message) {
            var data = JSON.parse(message.data);
            _this.emit(ClientEvent.Message, data);
            // const { channel, result } = data
            // let {id} = data
            // if (id) {
            //   const action = id.split(':')[1]
            //   if (action === 'sub') {
            //     const [channelId] = result
            //     this.channelIdToId.set(channelId, id)
            //     id += ':ack'
            //   }
            // } else if (channel) {
            //   id = this.channelIdToId.get(channel)
            // }
            // else {
            //   console.error('Unknown event:', data)
            //   return
            // }
            // const [clientName] = id.split(':')
            // const event = {...data, id}
            // const client = this[clientName]
            // if (client) {
            //   client.ws.handleEvent(event)
            // } else {
            //   console.error('Unknown client:', clientName)
            // }
        };
    };
    WsClient.prototype.disconnect = function () {
        this.socket.close();
        this.emit(ClientEvent.Disconnected);
    };
    WsClient.prototype.handleEvent = function (event) {
        this.emit(event.id, event);
    };
    WsClient.prototype.send = function (id, method, params) {
        try {
            this.socket.send(JSON.stringify({
                id: id,
                method: method,
                params: params,
            }));
        }
        catch (e) {
            console.error(e.message);
        }
    };
    WsClient.prototype.subscribe = function (id, channelId) {
        try {
            this.send(id, 'subscribe', { channels: [channelId] });
        }
        catch (e) {
            console.error(e.message);
        }
    };
    WsClient.prototype.unsubscribe = function (id, channelId) {
        try {
            this.socket.send(JSON.stringify({
                id: id,
                method: 'unsubscribe',
                params: { 'channels': [channelId] }
            }));
        }
        catch (e) {
            console.error(e.message);
        }
    };
    WsClient.prototype.subscribeMarketStats = function () {
        var channelId = 'market_stats';
        var id = "market:sub:" + channelId;
        this.subscribe(id, channelId);
    };
    WsClient.prototype.unsubscribeMarketStats = function () {
        var channelId = 'market_stats';
        var id = "market:unsub:" + channelId;
        this.unsubscribe(id, channelId);
    };
    return WsClient;
}(events_1.EventEmitter));
exports.WsClient = WsClient;
