// tradehub-api-js/build/main/lib/utils/address.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NEOAddress = exports.SWTHAddress = exports.wifEncodePrivateKey = exports.randomMnemonic = exports.stringOrBufferToBuffer = void 0;
var BIP32 = __importStar(require("bip32"));
var BIP39 = __importStar(require("bip39"));
var bip44_constants_1 = __importDefault(require("bip44-constants"));
var wallet_1 = require("./wallet");
var secp256r1 = __importStar(require("secp256r1"));
var secp256k1 = __importStar(require("secp256k1"));
var sha_js_1 = require("sha.js");
var bech32 = __importStar(require("bech32"));
var wif = __importStar(require("wif"));
var ripemd160 = __importStar(require("ripemd160"));
var Base58Check = __importStar(require("base58check"));
var config_1 = require("../config");
var BIP44_PURPOSE = 44;
var NEO_NAME = 'NEO';
var NEO_TOKEN_BIP44 = bip44_constants_1.default.filter(function (coin) { return coin[1] === NEO_NAME; })[0];
var NEO_COIN_TYPE = NEO_TOKEN_BIP44[0] - 0x80000000;
var SWTH_COIN_TYPE = 118;
/**
 * Convenience function to coalesce a string/buffer into a buffer
 *
 * @param stringOrBuffer a string or buffer type
 * @param encoding BufferEncoding from Buffer
 * @param force option to return an empty buffer regardless of input
 */
exports.stringOrBufferToBuffer = function (stringOrBuffer, encoding, force) {
    if (encoding === void 0) { encoding = 'hex'; }
    if (force === void 0) { force = false; }
    if (typeof stringOrBuffer === 'string') {
        return Buffer.from(stringOrBuffer, encoding);
    }
    if (stringOrBuffer instanceof Buffer) {
        return stringOrBuffer;
    }
    // not a string nor buffer
    // e.g. null/undefined
    if (force) {
        return Buffer.alloc(0);
    }
    // if not forcing to return an empty buffer, return null
    return null;
};
exports.randomMnemonic = function () {
    return BIP39.generateMnemonic();
};
exports.wifEncodePrivateKey = function (privateKey, iter) {
    if (iter === void 0) { iter = 128; }
    var privateKeyBuf = exports.stringOrBufferToBuffer(privateKey);
    return wif.encode(iter, privateKeyBuf, true);
};
exports.SWTHAddress = {
    coinType: function () {
        return SWTH_COIN_TYPE;
    },
    publicKeyToScriptHash: function (publicKey) {
        var sha256Hash = new sha_js_1.sha256().update(publicKey).digest();
        var ripemdHash = new ripemd160().update(sha256Hash).digest('hex');
        return ripemdHash;
    },
    publicKeyToAddress: function (publicKey, opts) {
        var scriptHash = exports.SWTHAddress.publicKeyToScriptHash(publicKey);
        var words = bech32.toWords(Buffer.from(scriptHash));
        var addressPrefix = config_1.getBech32Prefix(opts.network, opts.type);
        var addressBech32 = bech32.encode(addressPrefix, words);
        return addressBech32;
    },
    encodePublicKey: function () {
        throw new Error('SWTH public keys do not compress');
    },
    mnemonicToPrivateKey: function (mnemonic, account) {
        if (account === void 0) { account = 0; }
        var coinType = exports.SWTHAddress.coinType();
        var path = new wallet_1.BIP44(BIP44_PURPOSE, coinType).pathString(account);
        var seed = BIP39.mnemonicToSeedSync(mnemonic);
        var masterKey = BIP32.fromSeed(seed);
        var hardenedDerivation = masterKey.derivePath(path);
        var privateKey = hardenedDerivation.privateKey;
        if (!privateKey)
            throw new Error('Private key derivation from mnemonic failed');
        return privateKey;
    },
    generateAddress: function (mnemonic, account) {
        if (account === void 0) { account = 0; }
        var privateKey = exports.SWTHAddress.mnemonicToPrivateKey(mnemonic, account);
        var publicKeyUint8Array = secp256k1.publicKeyCreate(privateKey, true);
        var publicKey = Buffer.from(publicKeyUint8Array);
        var address = exports.SWTHAddress.publicKeyToAddress(publicKey);
        return address;
    },
};
exports.NEOAddress = {
    coinType: function () {
        return NEO_COIN_TYPE;
    },
    publicKeyToScriptHash: function (publicKey) {
        var encodedPublicKey = exports.NEOAddress.encodePublicKey(publicKey);
        var addressScript = Buffer.concat([
            Buffer.from([0x21]),
            encodedPublicKey,
            Buffer.from([0xac]),
        ]);
        var sha256Hash = new sha_js_1.sha256().update(addressScript).digest();
        var ripemdHash = new ripemd160().update(sha256Hash).digest('hex');
        return ripemdHash;
    },
    publicKeyToAddress: function (publicKey) {
        var addressScript = exports.NEOAddress.publicKeyToScriptHash(publicKey);
        var address = Base58Check.encode(addressScript, '17');
        return address;
    },
    encodePublicKey: function (unencodedPublicKey) {
        var unencPubKeyBuf = exports.stringOrBufferToBuffer(unencodedPublicKey);
        if (unencPubKeyBuf.length <= 33) {
            // length indicates already encoded
            return unencPubKeyBuf;
        }
        var pointXHex = unencPubKeyBuf.slice(1, 33);
        var pointYEven = unencPubKeyBuf[unencPubKeyBuf.length - 1] % 2 === 0;
        var compressedPublicKey = Buffer.concat([
            Buffer.from([pointYEven ? 0x02 : 0x03]),
            pointXHex,
        ]);
        return compressedPublicKey;
    },
    mnemonicToPrivateKey: function (mnemonic, account) {
        if (account === void 0) { account = 0; }
        var coinType = exports.NEOAddress.coinType();
        var path = new wallet_1.BIP44(BIP44_PURPOSE, coinType).pathString(account);
        var seed = BIP39.mnemonicToSeedSync(mnemonic);
        var masterKey = BIP32.fromSeed(seed);
        var hardenedDerivation = masterKey.derivePath(path);
        var privateKey = hardenedDerivation.privateKey;
        if (!privateKey)
            throw new Error('Private key derivation from mnemonic failed');
        return privateKey;
    },
    generateAddress: function (mnemonic, account) {
        if (account === void 0) { account = 0; }
        var privateKey = exports.NEOAddress.mnemonicToPrivateKey(mnemonic, account);
        var publicKeyUint8Array = secp256r1.publicKeyCreate(privateKey, true);
        var compressedPublicKey = Buffer.from(publicKeyUint8Array);
        var address = exports.NEOAddress.publicKeyToAddress(compressedPublicKey);
        return address;
    },
};
