// tradehub-api-js/build/module/lib/api/ws_wrapper.js
/*
  WsWrapper is a wrapper that helps to manage the socket connection with the server.
  Manages IDs for messages

  Methods:
    connect
    disconnect
    request
    subscribe
    unsubscribe
*/
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
import { getNetwork } from '../config';
/**
 * @deprecated
 * use WSConnector
 */
var WsWrapper = /** @class */ (function () {
    function WsWrapper(net, onMsgCallback) {
        this.isConnected = false;
        this.getBaseUrls = function (network) {
            var net = getNetwork(network);
            return net.WS_URL;
        };
        this.serverWsUrl = this.getBaseUrls(net);
        this.onMsgCallback = onMsgCallback;
    }
    WsWrapper.prototype.connect = function () {
        var _this = this;
        this.socket = new WebSocket(this.serverWsUrl);
        // Config socket
        this.socket.onopen = function () {
            _this.isConnected = true;
            console.log('socket is connected');
            _this.onMsgCallback("connected");
        };
        this.socket.onmessage = function (msg) {
            _this.onMsgCallback(msg);
        };
        this.socket.onclose = function () {
            _this.isConnected = false;
            console.log('socket off');
        };
    };
    WsWrapper.prototype.disconnect = function () {
        console.log("closing socket...");
        this.socket.close();
    };
    WsWrapper.prototype.checkIsConnected = function () {
        return this.isConnected;
    };
    // WS Get requests
    WsWrapper.prototype.wsGetOrderHistory = function (msgId, params) {
        try {
            var msg = JSON.stringify({
                id: msgId,
                method: 'get_order_history',
                params: { market: params.market, address: params.address }
            });
            this.socket.send(msg);
        }
        catch (e) {
            console.log(e.message);
        }
    };
    WsWrapper.prototype.wsGetRecentTrades = function (msgId, p) {
        try {
            var msg = JSON.stringify({
                id: msgId,
                method: 'get_recent_trades',
                params: { market: p.market }
            });
            this.socket.send(msg);
        }
        catch (e) {
            console.log(e.message);
        }
    };
    WsWrapper.prototype.wsGetCandlesticks = function (msgId, p) {
        try {
            var msg = JSON.stringify({
                id: msgId,
                method: 'get_candlesticks',
                params: {
                    market: p.market, resolution: p.resolution,
                    from: p.from, to: p.to
                }
            });
            this.socket.send(msg);
        }
        catch (e) {
            console.log(e.message);
        }
    };
    WsWrapper.prototype.wsGetOpenOrders = function (msgId, params) {
        try {
            var msg = JSON.stringify({
                id: msgId,
                method: 'get_open_orders',
                params: { market: params.market, address: params.address }
            });
            this.socket.send(msg);
        }
        catch (e) {
            console.log(e.message);
        }
    };
    WsWrapper.prototype.wsGetAccountTrades = function (msgId, params) {
        try {
            var msg = JSON.stringify({
                id: msgId,
                method: 'get_account_trades',
                params: params,
            });
            this.socket.send(msg);
        }
        catch (e) {
            console.log(e.message);
        }
    };
    WsWrapper.prototype.wsGetMarketStats = function (msgId, params) {
        try {
            var msg = JSON.stringify({
                id: msgId,
                method: 'get_market_stats',
                params: params,
            });
            this.socket.send(msg);
        }
        catch (e) {
            console.log(e.message);
        }
    };
    WsWrapper.prototype.wsGetLeverages = function (msgId, params) {
        try {
            var msg = JSON.stringify({
                id: msgId,
                method: 'get_leverages',
                params: params,
            });
            this.socket.send(msg);
        }
        catch (e) {
            console.log(e.message);
        }
    };
    WsWrapper.prototype.wsGetOpenPositions = function (msgId, params) {
        try {
            var msg = JSON.stringify({
                id: msgId,
                method: 'get_open_positions',
                params: params,
            });
            this.socket.send(msg);
        }
        catch (e) {
            console.log(e.message);
        }
    };
    WsWrapper.prototype.wsGetClosedPositions = function (msgId, params) {
        try {
            var msg = JSON.stringify({
                id: msgId,
                method: 'get_closed_positions',
                params: params,
            });
            this.socket.send(msg);
        }
        catch (e) {
            console.log(e.message);
        }
    };
    // WS Subscriptions
    WsWrapper.prototype.subscribe = function (msgId, params) {
        var _this = this;
        try {
            var channelIds = params.map(function (p) { return _this.generateChannelId(p); });
            // console.log("Subscribing to " + msgId)
            var msg = JSON.stringify({
                id: msgId,
                method: 'subscribe',
                params: { "channels": __spreadArrays(channelIds) }
            });
            this.socket.send(msg);
        }
        catch (e) {
            console.log(e.message);
        }
    };
    WsWrapper.prototype.unsubscribe = function (msgId, params) {
        var _this = this;
        try {
            var channelIds = params.map(function (p) { return _this.generateChannelId(p); });
            // console.log("Unsubscribing to " + channelIds)
            var msg = JSON.stringify({
                id: msgId,
                method: 'unsubscribe',
                params: { "channels": __spreadArrays(channelIds) }
            });
            this.socket.send(msg);
        }
        catch (e) {
            // console.log(e.message)
        }
    };
    WsWrapper.prototype.generateChannelId = function (p) {
        switch (p.channel) {
            case 'candlesticks': {
                var _a = p, channel = _a.channel, market = _a.market, resolution = _a.resolution;
                return [channel, market, resolution].join('.');
            }
            case 'books': {
                var _b = p, channel = _b.channel, market = _b.market;
                return [channel, market].join('.');
            }
            case 'recent_trades': {
                var _c = p, channel = _c.channel, market = _c.market;
                return [channel, market].join('.');
            }
            case 'orders': {
                var _d = p, channel = _d.channel, address = _d.address;
                return [channel, address].join('.');
            }
            case 'orders_by_market': {
                var _e = p, channel = _e.channel, market = _e.market, address = _e.address;
                return [channel, market, address].join('.');
            }
            case 'balances': {
                var _f = p, channel = _f.channel, address = _f.address;
                return [channel, address].join('.');
            }
            case 'account_trades': {
                var _g = p, channel = _g.channel, address = _g.address;
                return [channel, address].join('.');
            }
            case 'account_trades_by_market': {
                var _h = p, channel = _h.channel, market = _h.market, address = _h.address;
                return [channel, market, address].join('.');
            }
            case 'market_stats': {
                var channel = p.channel;
                return [channel].join('.');
            }
            case 'market_stats_by_market': {
                var _j = p, channel = _j.channel, market = _j.market;
                return [channel, market].join('.');
            }
            case 'leverages': {
                var _k = p, channel = _k.channel, address = _k.address;
                return [channel, address].join('.');
            }
            case 'leverages_by_market': {
                var _l = p, channel = _l.channel, market = _l.market, address = _l.address;
                return [channel, market, address].join('.');
            }
            case 'positions': {
                var _m = p, channel = _m.channel, address = _m.address;
                return [channel, address].join('.');
            }
            case 'positions_by_market': {
                var _o = p, channel = _o.channel, market = _o.market, address = _o.address;
                return [channel, market, address].join('.');
            }
            default:
                throw new Error("Invalid subscription");
        }
    };
    WsWrapper.parseChannelId = function (rawChannelId) {
        var _a = rawChannelId.split("."), channel = _a[0], market = _a[1], resolution = _a[2], address = _a[3];
        switch (channel) {
            case 'candlesticks':
                return {
                    channel: channel,
                    market: market,
                    resolution: resolution,
                };
            case 'books':
                return {
                    channel: channel,
                    market: market,
                };
            case 'recent_trades':
                return {
                    channel: channel,
                    market: market,
                };
            case 'orders':
                return {
                    channel: channel,
                    address: address,
                };
            case 'orders_by_market':
                return {
                    channel: channel,
                    market: market,
                    address: address,
                };
            case 'balances':
                return {
                    channel: channel,
                    address: address,
                };
            case 'account_trades':
                return {
                    channel: channel,
                    address: address,
                };
            case 'account_trades_by_market':
                return {
                    channel: channel,
                    market: market,
                    address: address,
                };
            case 'market_stats':
                return {
                    channel: channel,
                };
            case 'market_stats_by_market':
                return {
                    channel: channel,
                    market: market,
                };
            case 'leverages':
                return {
                    channel: channel,
                    address: address,
                };
            case 'leverages_by_market':
                return {
                    channel: channel,
                    market: market,
                    address: address,
                };
            case 'positions':
                return {
                    channel: channel,
                    address: address,
                };
            case 'positions_by_market':
                return {
                    channel: channel,
                    market: market,
                    address: address,
                };
            default:
                throw new Error("Error parsing channelId");
        }
    };
    return WsWrapper;
}());
export { WsWrapper };
