// tradehub-api-js/build/module/lib/types.js
// Orders
export var CREATE_ORDER_MSG_TYPE = 'order/MsgCreateOrder';
export var CANCEL_ORDER_MSG_TYPE = 'order/MsgCancelOrder';
export var CANCEL_ALL_MSG_TYPE = 'order/MsgCancelAll';
export var EDIT_ORDER_MSG_TYPE = 'order/MsgEditOrder';
export var CREATE_MARKET_MSG_TYPE = 'market/MsgCreateMarket';
export var UPDATE_MARKET_MSG_TYPE = 'market/MsgUpdateMarket';
export var INITIATE_SETTLEMENT_MSG_TYPE = 'broker/MsgInitiateSettlement';
// Positions
export var SET_LEVERAGE_MSG_TYPE = 'leverage/MsgSetLeverage';
export var EDIT_MARGIN_MSG_TYPE = 'position/MsgSetMargin';
// Tokens
export var MINT_TOKEN_MSG_TYPE = 'coin/MsgMintToken';
export var CREATE_TOKEN_MSG_TYPE = 'coin/MsgCreateToken';
export var CREATE_WITHDRAWAL_TYPE = 'coin/MsgWithdraw';
export var SEND_TOKENS_TYPE = 'cosmos-sdk/MsgSend';
// Oracle
export var CREATE_ORACLE_TYPE = 'oracle/MsgCreateOracle';
export var CREATE_VOTE_TYPE = 'oracle/MsgCreateVote';
// Staking
export var CREATE_VALIDATOR_MSG_TYPE = 'cosmos-sdk/MsgCreateValidator';
export var DELEGATE_TOKENS_MSG_TYPE = 'cosmos-sdk/MsgDelegate';
export var BEGIN_UNBONDING_TOKENS_MSG_TYPE = 'cosmos-sdk/MsgUndelegate';
export var BEGIN_REDELEGATING_TOKENS_MSG_TYPE = 'cosmos-sdk/MsgBeginRedelegate';
export var WITHDRAW_DELEGATOR_REWARDS_MSG_TYPE = 'cosmos-sdk/MsgWithdrawDelegationReward';
// Accounts
export var CREATE_SUB_ACCOUNT_MSG_TYPE = 'subaccount/MsgCreateSubAccountV1';
export var ACTIVATE_SUB_ACCOUNT_MSG_TYPE = 'subaccount/MsgActivateSubAccountV1';
// Profile
export var UPDATE_PROFILE_MSG_TYPE = 'profile/MsgUpdateProfile';
// Gov
export var SUBMIT_PROPOSAL_TYPE = 'cosmos-sdk/MsgSubmitProposal';
export var DEPOSIT_PROPOSAL_TYPE = 'cosmos-sdk/MsgDeposit';
export var VOTE_PROPOSAL_TYPE = 'cosmos-sdk/MsgVote';
// AMM
export var ADD_LIQUIDITY_MSG_TYPE = 'liquiditypool/AddLiquidity';
export var REMOVE_LIQUIDITY_MSG_TYPE = 'liquiditypool/RemoveLiquidity';
export var CREATE_POOL_MSG_TYPE = 'liquiditypool/CreatePool';
export var CREATE_POOL_WITH_LIQUIDITY_MSG_TYPE = 'liquiditypool/CreatePoolWithLiquidity';
export var LINK_POOL_MSG_TYPE = 'liquiditypool/LinkPool';
export var UNLINK_POOL_MSG_TYPE = 'liquiditypool/UnlinkPool';
export var SET_REWARDS_WEIGHTS_MSG_TYPE = 'liquiditypool/SetRewardsWeights';
export var SET_REWARD_CURVE_MSG_TYPE = 'liquiditypool/SetRewardCurve';
export var SET_COMMITMENT_CURVE_MSG_TYPE = 'liquiditypool/SetCommitmentCurve';
export var STAKE_POOL_TOKEN_MSG_TYPE = 'liquiditypool/StakePoolToken';
export var UNSTAKE_POOL_TOKEN_MSG_TYPE = 'liquiditypool/UnstakePoolToken';
export var CLAIM_POOL_REWARDS_MSG_TYPE = 'liquiditypool/ClaimPoolRewards';
export var LINK_POOL_PROPOSAL_TYPE = 'liquiditypool/LinkPoolProposal';
export var SET_REWARD_CURVE_PROPOSAL_TYPE = 'liquiditypool/SetRewardCurveProposal';
export var SET_REWARDS_WEIGHT_PROPOSAL_TYPE = 'liquiditypool/SetRewardsWeightsProposal';
export var SET_COMMITMENT_CURVE_PROPOSAL_TYPE = 'liquiditypool/SetCommitmentCurveProposal';
export var CHANGE_SWAP_FEE_PROPOSAL_TYPE = 'liquiditypool/ChangeSwapFeeProposal';
export var CHANGE_NUM_QUOTES_PROPOSAL_TYPE = 'liquiditypool/ChangeNumQuotesProposal';
// CDP
export var CREATE_VAULT_TYPE_MSG_TYPE = 'collateralizeddebtposition/CreateVaultType';
export var ADD_COLLATERAL_MSG_TYPE = 'collateralizeddebtposition/AddCollateral';
export var REMOVE_COLLATERAL_MSG_TYPE = 'collateralizeddebtposition/RemoveCollateral';
export var ADD_DEBT_MSG_TYPE = 'collateralizeddebtposition/AddDebt';
export var REMOVE_DEBT_MSG_TYPE = 'collateralizeddebtposition/RemoveDebt';
// Fee
export var SET_MESSAGE_FEE_TYPE = 'fee/SetMsgFee';
export var SET_MESSAGE_FEE_PROPOSAL_TYPE = 'fee/SetMsgFeeProposal';
export var Network;
(function (Network) {
    Network["LocalHost"] = "LOCALHOST";
    Network["TestNet"] = "TESTNET";
    Network["MainNet"] = "MAINNET";
    Network["DevNet"] = "DEVNET";
})(Network || (Network = {}));
export var FEE_TYPES = {
    'order/MsgCreateOrder': 'create_order',
    'liquiditypool/ClaimPoolRewards': 'claim_pool_rewards',
    'oracle/MsgCreateOracle': 'create_oracle_vote',
    'liquiditypool/CreatePool': 'create_pool',
    'liquiditypool/StakePoolToken': 'stake_pool_token',
    'liquiditypool/UnstakePoolToken': 'unstake_pool_token',
    Default: 'default_fee'
};
