// tradehub-api-js/build/main/lib/containers/Transaction.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transaction = void 0;
var bignumber_js_1 = require("bignumber.js");
var config_1 = require("../config");
var StdSignDoc_1 = require("./StdSignDoc");
var Transaction = /** @class */ (function () {
    function Transaction(msgs, signatures, options) {
        if (options === void 0) { options = {}; }
        var fee = options.fee || new StdSignDoc_1.Fee([{ denom: "swth", amount: (new bignumber_js_1.BigNumber(msgs.length)).shiftedBy(8).toString() }], config_1.CONFIG.DEFAULT_GAS);
        this.mode = options.mode || 'block';
        this.tx = {
            fee: fee,
            msg: msgs,
            signatures: signatures,
        };
        if (options.memo) {
            this.tx = __assign(__assign({}, this.tx), { memo: options.memo });
        }
    }
    return Transaction;
}());
exports.Transaction = Transaction;
