// tradehub-api-js/build/module/lib/websocket/channel.js
import { WSChannel, } from './types';
export var generateChannelId = function (params) {
    switch (params.channel) {
        case WSChannel.candlesticks: {
            var _a = params, channel = _a.channel, market = _a.market, resolution = _a.resolution;
            return [channel, market, resolution].join('.');
        }
        case WSChannel.books: {
            var _b = params, channel = _b.channel, market = _b.market;
            return [channel, market].join('.');
        }
        case WSChannel.recent_trades: {
            var _c = params, channel = _c.channel, market = _c.market;
            return [channel, market].join('.');
        }
        case WSChannel.orders: {
            var _d = params, channel = _d.channel, address = _d.address;
            return [channel, address].join('.');
        }
        case WSChannel.orders_by_market: {
            var _e = params, channel = _e.channel, market = _e.market, address = _e.address;
            return [channel, market, address].join('.');
        }
        case WSChannel.balances: {
            var _f = params, channel = _f.channel, address = _f.address;
            return [channel, address].join('.');
        }
        case WSChannel.account_trades: {
            var _g = params, channel = _g.channel, address = _g.address;
            return [channel, address].join('.');
        }
        case WSChannel.account_trades_by_market: {
            var _h = params, channel = _h.channel, market = _h.market, address = _h.address;
            return [channel, market, address].join('.');
        }
        case WSChannel.market_stats: {
            var channel = params.channel;
            return [channel].join('.');
        }
        case WSChannel.market_stats_by_market: {
            var _j = params, channel = _j.channel, market = _j.market;
            return [channel, market].join('.');
        }
        case WSChannel.leverages: {
            var _k = params, channel = _k.channel, address = _k.address;
            return [channel, address].join('.');
        }
        case WSChannel.leverages_by_market: {
            var _l = params, channel = _l.channel, market = _l.market, address = _l.address;
            return [channel, market, address].join('.');
        }
        case WSChannel.positions: {
            var _m = params, channel = _m.channel, address = _m.address;
            return [channel, address].join('.');
        }
        case WSChannel.positions_by_market: {
            var _o = params, channel = _o.channel, market = _o.market, address = _o.address;
            return [channel, market, address].join('.');
        }
        default:
            throw new Error("invalid subscription channel: " + params.channel);
    }
};
export var parseChannelId = function (rawChannelId) {
    var _a = rawChannelId.split('.'), channel = _a[0], market = _a[1], resolution = _a[2], address = _a[3];
    switch (channel) {
        case WSChannel.candlesticks:
            return {
                channel: channel,
                market: market,
                resolution: resolution,
            };
        case WSChannel.books:
            return {
                channel: channel,
                market: market,
            };
        case WSChannel.recent_trades:
            return {
                channel: channel,
                market: market,
            };
        case WSChannel.orders:
            return {
                channel: channel,
                address: address,
            };
        case WSChannel.orders_by_market:
            return {
                channel: channel,
                market: market,
                address: address,
            };
        case WSChannel.balances:
            return {
                channel: channel,
                address: address,
            };
        case WSChannel.account_trades:
            return {
                channel: channel,
                address: address,
            };
        case WSChannel.account_trades_by_market:
            return {
                channel: channel,
                market: market,
                address: address,
            };
        case WSChannel.market_stats:
            return {
                channel: channel,
            };
        case WSChannel.market_stats_by_market:
            return {
                channel: channel,
                market: market,
            };
        case WSChannel.leverages:
            return {
                channel: channel,
                address: address,
            };
        case WSChannel.leverages_by_market:
            return {
                channel: channel,
                market: market,
                address: address,
            };
        case WSChannel.positions:
            return {
                channel: channel,
                address: address,
            };
        case WSChannel.positions_by_market:
            return {
                channel: channel,
                market: market,
                address: address,
            };
        default:
            throw new Error('Error parsing channelId');
    }
};
