// tradehub-api-js/build/main/lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FEE_TYPES = exports.Network = exports.SET_MESSAGE_FEE_PROPOSAL_TYPE = exports.SET_MESSAGE_FEE_TYPE = exports.REMOVE_DEBT_MSG_TYPE = exports.ADD_DEBT_MSG_TYPE = exports.REMOVE_COLLATERAL_MSG_TYPE = exports.ADD_COLLATERAL_MSG_TYPE = exports.CREATE_VAULT_TYPE_MSG_TYPE = exports.CHANGE_NUM_QUOTES_PROPOSAL_TYPE = exports.CHANGE_SWAP_FEE_PROPOSAL_TYPE = exports.SET_COMMITMENT_CURVE_PROPOSAL_TYPE = exports.SET_REWARDS_WEIGHT_PROPOSAL_TYPE = exports.SET_REWARD_CURVE_PROPOSAL_TYPE = exports.LINK_POOL_PROPOSAL_TYPE = exports.CLAIM_POOL_REWARDS_MSG_TYPE = exports.UNSTAKE_POOL_TOKEN_MSG_TYPE = exports.STAKE_POOL_TOKEN_MSG_TYPE = exports.SET_COMMITMENT_CURVE_MSG_TYPE = exports.SET_REWARD_CURVE_MSG_TYPE = exports.SET_REWARDS_WEIGHTS_MSG_TYPE = exports.UNLINK_POOL_MSG_TYPE = exports.LINK_POOL_MSG_TYPE = exports.CREATE_POOL_WITH_LIQUIDITY_MSG_TYPE = exports.CREATE_POOL_MSG_TYPE = exports.REMOVE_LIQUIDITY_MSG_TYPE = exports.ADD_LIQUIDITY_MSG_TYPE = exports.VOTE_PROPOSAL_TYPE = exports.DEPOSIT_PROPOSAL_TYPE = exports.SUBMIT_PROPOSAL_TYPE = exports.UPDATE_PROFILE_MSG_TYPE = exports.ACTIVATE_SUB_ACCOUNT_MSG_TYPE = exports.CREATE_SUB_ACCOUNT_MSG_TYPE = exports.WITHDRAW_DELEGATOR_REWARDS_MSG_TYPE = exports.BEGIN_REDELEGATING_TOKENS_MSG_TYPE = exports.BEGIN_UNBONDING_TOKENS_MSG_TYPE = exports.DELEGATE_TOKENS_MSG_TYPE = exports.CREATE_VALIDATOR_MSG_TYPE = exports.CREATE_VOTE_TYPE = exports.CREATE_ORACLE_TYPE = exports.SEND_TOKENS_TYPE = exports.CREATE_WITHDRAWAL_TYPE = exports.CREATE_TOKEN_MSG_TYPE = exports.MINT_TOKEN_MSG_TYPE = exports.EDIT_MARGIN_MSG_TYPE = exports.SET_LEVERAGE_MSG_TYPE = exports.INITIATE_SETTLEMENT_MSG_TYPE = exports.UPDATE_MARKET_MSG_TYPE = exports.CREATE_MARKET_MSG_TYPE = exports.EDIT_ORDER_MSG_TYPE = exports.CANCEL_ALL_MSG_TYPE = exports.CANCEL_ORDER_MSG_TYPE = exports.CREATE_ORDER_MSG_TYPE = void 0;
// Orders
exports.CREATE_ORDER_MSG_TYPE = 'order/MsgCreateOrder';
exports.CANCEL_ORDER_MSG_TYPE = 'order/MsgCancelOrder';
exports.CANCEL_ALL_MSG_TYPE = 'order/MsgCancelAll';
exports.EDIT_ORDER_MSG_TYPE = 'order/MsgEditOrder';
exports.CREATE_MARKET_MSG_TYPE = 'market/MsgCreateMarket';
exports.UPDATE_MARKET_MSG_TYPE = 'market/MsgUpdateMarket';
exports.INITIATE_SETTLEMENT_MSG_TYPE = 'broker/MsgInitiateSettlement';
// Positions
exports.SET_LEVERAGE_MSG_TYPE = 'leverage/MsgSetLeverage';
exports.EDIT_MARGIN_MSG_TYPE = 'position/MsgSetMargin';
// Tokens
exports.MINT_TOKEN_MSG_TYPE = 'coin/MsgMintToken';
exports.CREATE_TOKEN_MSG_TYPE = 'coin/MsgCreateToken';
exports.CREATE_WITHDRAWAL_TYPE = 'coin/MsgWithdraw';
exports.SEND_TOKENS_TYPE = 'cosmos-sdk/MsgSend';
// Oracle
exports.CREATE_ORACLE_TYPE = 'oracle/MsgCreateOracle';
exports.CREATE_VOTE_TYPE = 'oracle/MsgCreateVote';
// Staking
exports.CREATE_VALIDATOR_MSG_TYPE = 'cosmos-sdk/MsgCreateValidator';
exports.DELEGATE_TOKENS_MSG_TYPE = 'cosmos-sdk/MsgDelegate';
exports.BEGIN_UNBONDING_TOKENS_MSG_TYPE = 'cosmos-sdk/MsgUndelegate';
exports.BEGIN_REDELEGATING_TOKENS_MSG_TYPE = 'cosmos-sdk/MsgBeginRedelegate';
exports.WITHDRAW_DELEGATOR_REWARDS_MSG_TYPE = 'cosmos-sdk/MsgWithdrawDelegationReward';
// Accounts
exports.CREATE_SUB_ACCOUNT_MSG_TYPE = 'subaccount/MsgCreateSubAccountV1';
exports.ACTIVATE_SUB_ACCOUNT_MSG_TYPE = 'subaccount/MsgActivateSubAccountV1';
// Profile
exports.UPDATE_PROFILE_MSG_TYPE = 'profile/MsgUpdateProfile';
// Gov
exports.SUBMIT_PROPOSAL_TYPE = 'cosmos-sdk/MsgSubmitProposal';
exports.DEPOSIT_PROPOSAL_TYPE = 'cosmos-sdk/MsgDeposit';
exports.VOTE_PROPOSAL_TYPE = 'cosmos-sdk/MsgVote';
// AMM
exports.ADD_LIQUIDITY_MSG_TYPE = 'liquiditypool/AddLiquidity';
exports.REMOVE_LIQUIDITY_MSG_TYPE = 'liquiditypool/RemoveLiquidity';
exports.CREATE_POOL_MSG_TYPE = 'liquiditypool/CreatePool';
exports.CREATE_POOL_WITH_LIQUIDITY_MSG_TYPE = 'liquiditypool/CreatePoolWithLiquidity';
exports.LINK_POOL_MSG_TYPE = 'liquiditypool/LinkPool';
exports.UNLINK_POOL_MSG_TYPE = 'liquiditypool/UnlinkPool';
exports.SET_REWARDS_WEIGHTS_MSG_TYPE = 'liquiditypool/SetRewardsWeights';
exports.SET_REWARD_CURVE_MSG_TYPE = 'liquiditypool/SetRewardCurve';
exports.SET_COMMITMENT_CURVE_MSG_TYPE = 'liquiditypool/SetCommitmentCurve';
exports.STAKE_POOL_TOKEN_MSG_TYPE = 'liquiditypool/StakePoolToken';
exports.UNSTAKE_POOL_TOKEN_MSG_TYPE = 'liquiditypool/UnstakePoolToken';
exports.CLAIM_POOL_REWARDS_MSG_TYPE = 'liquiditypool/ClaimPoolRewards';
exports.LINK_POOL_PROPOSAL_TYPE = 'liquiditypool/LinkPoolProposal';
exports.SET_REWARD_CURVE_PROPOSAL_TYPE = 'liquiditypool/SetRewardCurveProposal';
exports.SET_REWARDS_WEIGHT_PROPOSAL_TYPE = 'liquiditypool/SetRewardsWeightsProposal';
exports.SET_COMMITMENT_CURVE_PROPOSAL_TYPE = 'liquiditypool/SetCommitmentCurveProposal';
exports.CHANGE_SWAP_FEE_PROPOSAL_TYPE = 'liquiditypool/ChangeSwapFeeProposal';
exports.CHANGE_NUM_QUOTES_PROPOSAL_TYPE = 'liquiditypool/ChangeNumQuotesProposal';
// CDP
exports.CREATE_VAULT_TYPE_MSG_TYPE = 'collateralizeddebtposition/CreateVaultType';
exports.ADD_COLLATERAL_MSG_TYPE = 'collateralizeddebtposition/AddCollateral';
exports.REMOVE_COLLATERAL_MSG_TYPE = 'collateralizeddebtposition/RemoveCollateral';
exports.ADD_DEBT_MSG_TYPE = 'collateralizeddebtposition/AddDebt';
exports.REMOVE_DEBT_MSG_TYPE = 'collateralizeddebtposition/RemoveDebt';
// Fee
exports.SET_MESSAGE_FEE_TYPE = 'fee/SetMsgFee';
exports.SET_MESSAGE_FEE_PROPOSAL_TYPE = 'fee/SetMsgFeeProposal';
var Network;
(function (Network) {
    Network["LocalHost"] = "LOCALHOST";
    Network["TestNet"] = "TESTNET";
    Network["MainNet"] = "MAINNET";
    Network["DevNet"] = "DEVNET";
})(Network = exports.Network || (exports.Network = {}));
exports.FEE_TYPES = {
    'order/MsgCreateOrder': 'create_order',
    'liquiditypool/ClaimPoolRewards': 'claim_pool_rewards',
    'oracle/MsgCreateOracle': 'create_oracle_vote',
    'liquiditypool/CreatePool': 'create_pool',
    'liquiditypool/StakePoolToken': 'stake_pool_token',
    'liquiditypool/UnstakePoolToken': 'unstake_pool_token',
    Default: 'default_fee'
};
