// tradehub-api-js/build/module/lib/containers/Transaction.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { BigNumber } from 'bignumber.js';
import { CONFIG } from '../config';
import { Fee } from './StdSignDoc';
var Transaction = /** @class */ (function () {
    function Transaction(msgs, signatures, options) {
        if (options === void 0) { options = {}; }
        var fee = options.fee || new Fee([{ denom: "swth", amount: (new BigNumber(msgs.length)).shiftedBy(8).toString() }], CONFIG.DEFAULT_GAS);
        this.mode = options.mode || 'block';
        this.tx = {
            fee: fee,
            msg: msgs,
            signatures: signatures,
        };
        if (options.memo) {
            this.tx = __assign(__assign({}, this.tx), { memo: options.memo });
        }
    }
    return Transaction;
}());
export { Transaction };
