// tradehub-api-js/build/main/lib/websocket/channel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseChannelId = exports.generateChannelId = void 0;
var types_1 = require("./types");
exports.generateChannelId = function (params) {
    switch (params.channel) {
        case types_1.WSChannel.candlesticks: {
            var _a = params, channel = _a.channel, market = _a.market, resolution = _a.resolution;
            return [channel, market, resolution].join('.');
        }
        case types_1.WSChannel.books: {
            var _b = params, channel = _b.channel, market = _b.market;
            return [channel, market].join('.');
        }
        case types_1.WSChannel.recent_trades: {
            var _c = params, channel = _c.channel, market = _c.market;
            return [channel, market].join('.');
        }
        case types_1.WSChannel.orders: {
            var _d = params, channel = _d.channel, address = _d.address;
            return [channel, address].join('.');
        }
        case types_1.WSChannel.orders_by_market: {
            var _e = params, channel = _e.channel, market = _e.market, address = _e.address;
            return [channel, market, address].join('.');
        }
        case types_1.WSChannel.balances: {
            var _f = params, channel = _f.channel, address = _f.address;
            return [channel, address].join('.');
        }
        case types_1.WSChannel.account_trades: {
            var _g = params, channel = _g.channel, address = _g.address;
            return [channel, address].join('.');
        }
        case types_1.WSChannel.account_trades_by_market: {
            var _h = params, channel = _h.channel, market = _h.market, address = _h.address;
            return [channel, market, address].join('.');
        }
        case types_1.WSChannel.market_stats: {
            var channel = params.channel;
            return [channel].join('.');
        }
        case types_1.WSChannel.market_stats_by_market: {
            var _j = params, channel = _j.channel, market = _j.market;
            return [channel, market].join('.');
        }
        case types_1.WSChannel.leverages: {
            var _k = params, channel = _k.channel, address = _k.address;
            return [channel, address].join('.');
        }
        case types_1.WSChannel.leverages_by_market: {
            var _l = params, channel = _l.channel, market = _l.market, address = _l.address;
            return [channel, market, address].join('.');
        }
        case types_1.WSChannel.positions: {
            var _m = params, channel = _m.channel, address = _m.address;
            return [channel, address].join('.');
        }
        case types_1.WSChannel.positions_by_market: {
            var _o = params, channel = _o.channel, market = _o.market, address = _o.address;
            return [channel, market, address].join('.');
        }
        default:
            throw new Error("invalid subscription channel: " + params.channel);
    }
};
exports.parseChannelId = function (rawChannelId) {
    var _a = rawChannelId.split('.'), channel = _a[0], market = _a[1], resolution = _a[2], address = _a[3];
    switch (channel) {
        case types_1.WSChannel.candlesticks:
            return {
                channel: channel,
                market: market,
                resolution: resolution,
            };
        case types_1.WSChannel.books:
            return {
                channel: channel,
                market: market,
            };
        case types_1.WSChannel.recent_trades:
            return {
                channel: channel,
                market: market,
            };
        case types_1.WSChannel.orders:
            return {
                channel: channel,
                address: address,
            };
        case types_1.WSChannel.orders_by_market:
            return {
                channel: channel,
                market: market,
                address: address,
            };
        case types_1.WSChannel.balances:
            return {
                channel: channel,
                address: address,
            };
        case types_1.WSChannel.account_trades:
            return {
                channel: channel,
                address: address,
            };
        case types_1.WSChannel.account_trades_by_market:
            return {
                channel: channel,
                market: market,
                address: address,
            };
        case types_1.WSChannel.market_stats:
            return {
                channel: channel,
            };
        case types_1.WSChannel.market_stats_by_market:
            return {
                channel: channel,
                market: market,
            };
        case types_1.WSChannel.leverages:
            return {
                channel: channel,
                address: address,
            };
        case types_1.WSChannel.leverages_by_market:
            return {
                channel: channel,
                market: market,
                address: address,
            };
        case types_1.WSChannel.positions:
            return {
                channel: channel,
                address: address,
            };
        case types_1.WSChannel.positions_by_market:
            return {
                channel: channel,
                market: market,
                address: address,
            };
        default:
            throw new Error('Error parsing channelId');
    }
};
