// tradehub-api-js/build/module/lib/utils/address.js
import * as BIP32 from 'bip32';
import * as BIP39 from 'bip39';
import bip44Constants from 'bip44-constants';
import { BIP44 } from './wallet';
import * as secp256r1 from 'secp256r1';
import * as secp256k1 from 'secp256k1';
import { sha256 } from 'sha.js';
import * as bech32 from 'bech32';
import * as wif from 'wif';
import * as ripemd160 from 'ripemd160';
import * as Base58Check from 'base58check';
import { getBech32Prefix } from '../config';
var BIP44_PURPOSE = 44;
var NEO_NAME = 'NEO';
var NEO_TOKEN_BIP44 = bip44Constants.filter(function (coin) { return coin[1] === NEO_NAME; })[0];
var NEO_COIN_TYPE = NEO_TOKEN_BIP44[0] - 0x80000000;
var SWTH_COIN_TYPE = 118;
/**
 * Convenience function to coalesce a string/buffer into a buffer
 *
 * @param stringOrBuffer a string or buffer type
 * @param encoding BufferEncoding from Buffer
 * @param force option to return an empty buffer regardless of input
 */
export var stringOrBufferToBuffer = function (stringOrBuffer, encoding, force) {
    if (encoding === void 0) { encoding = 'hex'; }
    if (force === void 0) { force = false; }
    if (typeof stringOrBuffer === 'string') {
        return Buffer.from(stringOrBuffer, encoding);
    }
    if (stringOrBuffer instanceof Buffer) {
        return stringOrBuffer;
    }
    // not a string nor buffer
    // e.g. null/undefined
    if (force) {
        return Buffer.alloc(0);
    }
    // if not forcing to return an empty buffer, return null
    return null;
};
export var randomMnemonic = function () {
    return BIP39.generateMnemonic();
};
export var wifEncodePrivateKey = function (privateKey, iter) {
    if (iter === void 0) { iter = 128; }
    var privateKeyBuf = stringOrBufferToBuffer(privateKey);
    return wif.encode(iter, privateKeyBuf, true);
};
export var SWTHAddress = {
    coinType: function () {
        return SWTH_COIN_TYPE;
    },
    publicKeyToScriptHash: function (publicKey) {
        var sha256Hash = new sha256().update(publicKey).digest();
        var ripemdHash = new ripemd160().update(sha256Hash).digest('hex');
        return ripemdHash;
    },
    publicKeyToAddress: function (publicKey, opts) {
        var scriptHash = SWTHAddress.publicKeyToScriptHash(publicKey);
        var words = bech32.toWords(Buffer.from(scriptHash));
        var addressPrefix = getBech32Prefix(opts.network, opts.type);
        var addressBech32 = bech32.encode(addressPrefix, words);
        return addressBech32;
    },
    encodePublicKey: function () {
        throw new Error('SWTH public keys do not compress');
    },
    mnemonicToPrivateKey: function (mnemonic, account) {
        if (account === void 0) { account = 0; }
        var coinType = SWTHAddress.coinType();
        var path = new BIP44(BIP44_PURPOSE, coinType).pathString(account);
        var seed = BIP39.mnemonicToSeedSync(mnemonic);
        var masterKey = BIP32.fromSeed(seed);
        var hardenedDerivation = masterKey.derivePath(path);
        var privateKey = hardenedDerivation.privateKey;
        if (!privateKey)
            throw new Error('Private key derivation from mnemonic failed');
        return privateKey;
    },
    generateAddress: function (mnemonic, account) {
        if (account === void 0) { account = 0; }
        var privateKey = SWTHAddress.mnemonicToPrivateKey(mnemonic, account);
        var publicKeyUint8Array = secp256k1.publicKeyCreate(privateKey, true);
        var publicKey = Buffer.from(publicKeyUint8Array);
        var address = SWTHAddress.publicKeyToAddress(publicKey);
        return address;
    },
};
export var NEOAddress = {
    coinType: function () {
        return NEO_COIN_TYPE;
    },
    publicKeyToScriptHash: function (publicKey) {
        var encodedPublicKey = NEOAddress.encodePublicKey(publicKey);
        var addressScript = Buffer.concat([
            Buffer.from([0x21]),
            encodedPublicKey,
            Buffer.from([0xac]),
        ]);
        var sha256Hash = new sha256().update(addressScript).digest();
        var ripemdHash = new ripemd160().update(sha256Hash).digest('hex');
        return ripemdHash;
    },
    publicKeyToAddress: function (publicKey) {
        var addressScript = NEOAddress.publicKeyToScriptHash(publicKey);
        var address = Base58Check.encode(addressScript, '17');
        return address;
    },
    encodePublicKey: function (unencodedPublicKey) {
        var unencPubKeyBuf = stringOrBufferToBuffer(unencodedPublicKey);
        if (unencPubKeyBuf.length <= 33) {
            // length indicates already encoded
            return unencPubKeyBuf;
        }
        var pointXHex = unencPubKeyBuf.slice(1, 33);
        var pointYEven = unencPubKeyBuf[unencPubKeyBuf.length - 1] % 2 === 0;
        var compressedPublicKey = Buffer.concat([
            Buffer.from([pointYEven ? 0x02 : 0x03]),
            pointXHex,
        ]);
        return compressedPublicKey;
    },
    mnemonicToPrivateKey: function (mnemonic, account) {
        if (account === void 0) { account = 0; }
        var coinType = NEOAddress.coinType();
        var path = new BIP44(BIP44_PURPOSE, coinType).pathString(account);
        var seed = BIP39.mnemonicToSeedSync(mnemonic);
        var masterKey = BIP32.fromSeed(seed);
        var hardenedDerivation = masterKey.derivePath(path);
        var privateKey = hardenedDerivation.privateKey;
        if (!privateKey)
            throw new Error('Private key derivation from mnemonic failed');
        return privateKey;
    },
    generateAddress: function (mnemonic, account) {
        if (account === void 0) { account = 0; }
        var privateKey = NEOAddress.mnemonicToPrivateKey(mnemonic, account);
        var publicKeyUint8Array = secp256r1.publicKeyCreate(privateKey, true);
        var compressedPublicKey = Buffer.from(publicKeyUint8Array);
        var address = NEOAddress.publicKeyToAddress(compressedPublicKey);
        return address;
    },
};
