// tradehub-api-js/build/module/lib/utils/wallet.js
// tslint:disable:max-classes-per-file
import bech32 from 'bech32';
import ripemd160 from 'ripemd160';
import secp256k1 from 'secp256k1';
import { sha256 } from 'sha.js';
var PrivKeySecp256k1 = /** @class */ (function () {
    function PrivKeySecp256k1(privKey) {
        this.privKey = privKey;
    }
    PrivKeySecp256k1.prototype.toBytes = function () {
        // return marshalBinaryBare(this) // stub marshalBinaryBare with Uint8Array
        return new Uint8Array(this.privKey);
    };
    PrivKeySecp256k1.prototype.toPubKey = function () {
        var pubKey = secp256k1.publicKeyCreate(Buffer.from(this.privKey), true);
        return new PubKeySecp256k1(Buffer.from(pubKey));
    };
    PrivKeySecp256k1.prototype.equals = function (privKey) {
        return this.toBytes().toString() === privKey.toBytes().toString();
    };
    PrivKeySecp256k1.prototype.sign = function (msg) {
        return secp256k1.ecdsaSign(Buffer.from(new sha256().update(msg).digest()), Buffer.from(this.privKey));
    };
    return PrivKeySecp256k1;
}());
export { PrivKeySecp256k1 };
var PubKeySecp256k1 = /** @class */ (function () {
    function PubKeySecp256k1(pubKey) {
        this.pubKey = pubKey;
    }
    PubKeySecp256k1.prototype.toBytes = function () {
        // return marshalBinaryBare(this)
        return new Uint8Array(this.pubKey);
    };
    PubKeySecp256k1.prototype.toAddress = function () {
        var hash = new sha256().update(this.pubKey).digest('latin1');
        hash = new ripemd160().update(hash, 'latin1').digest('hex');
        return new Address(Buffer.from(hash, 'hex'));
    };
    PubKeySecp256k1.prototype.equals = function (pubKey) {
        return this.toBytes().toString() === pubKey.toBytes().toString();
    };
    PubKeySecp256k1.prototype.verify = function (msg, sig) {
        return secp256k1.ecdsaVerify(Buffer.from(msg), Buffer.from(sig), Buffer.from(this.pubKey));
    };
    return PubKeySecp256k1;
}());
export { PubKeySecp256k1 };
var Address = /** @class */ (function () {
    function Address(address) {
        this.address = address;
    }
    Address.fromBech32 = function (prefix, bech32Addr) {
        var _a = bech32.decode(bech32Addr), b32Prefix = _a.prefix, words = _a.words;
        if (b32Prefix !== prefix) {
            throw new Error("Prefix doesn't match");
        }
        return new Address(bech32.fromWords(words));
    };
    Address.prototype.toBech32 = function (prefix) {
        var words = bech32.toWords(Buffer.from(this.address));
        return bech32.encode(prefix, words);
    };
    Address.prototype.toBytes = function () {
        return new Uint8Array(this.address);
    };
    return Address;
}());
export { Address };
var BIP44 = /** @class */ (function () {
    function BIP44(purpose, coinType, account) {
        if (purpose === void 0) { purpose = 44; }
        if (coinType === void 0) { coinType = 118; }
        if (account === void 0) { account = 0; }
        this.purpose = purpose;
        this.coinType = coinType;
        this.account = account;
    }
    BIP44.prototype.path = function (index, change) {
        if (change === void 0) { change = 0; }
        if (this.purpose !== parseInt(this.purpose.toString(), 10)) {
            throw new Error('Purpose should be integer');
        }
        if (this.coinType !== parseInt(this.coinType.toString(), 10)) {
            throw new Error('CoinType should be integer');
        }
        if (this.account !== parseInt(this.account.toString(), 10)) {
            throw new Error('Account should be integer');
        }
        if (change !== parseInt(change.toString(), 10)) {
            throw new Error('Change should be integer');
        }
        if (index !== parseInt(index.toString(), 10)) {
            throw new Error('Index should be integer');
        }
        return [this.purpose, this.coinType, this.account, change, index];
    };
    BIP44.prototype.pathString = function (index, change) {
        if (change === void 0) { change = 0; }
        var path = this.path(index, change);
        return "m/" + path[0] + "'/" + path[1] + "'/" + path[2] + "'/" + path[3] + "/" + path[4];
    };
    return BIP44;
}());
export { BIP44 };
export function getPathArray() {
    return new BIP44().path(0, 0);
}
export function getPath() {
    return new BIP44().pathString(0, 0);
}
